### ISSUE TITLE
cookie support?

### ISSUE THREAD

**Issue by Nervengift** [jonas-schievink/xml-rpc-rs#19]
I'm trying to use an xml-rpc API that uses cookies for authentication (send auth command -> receive cookie -> send cookie with following commands).  As far as I can see, 
`hyper::Client` does not directly support storing cookies and I see no easy way to work around your library here since I can not access the `Request` and `Response` directly.

Would handling session cookies be a feature you'd want in your lib? I'd try to submit a pull request then.

**Comment by jonas-schievink** [comment]
I'd love to make `Client`s API flexible enough to support this. Although I'm thinking it might make more sense to allow the caller to modify the request we send.

Since the only way to modify hyper requests is adding headers, a `call_with_headers` method that takes a `Headers` objects should be enough for now.

**Comment by Nervengift** [comment]
We'd also need a way of getting the headers from the response then.

I could just use `write_as_xml(...)`, do the call manually, save the cookie and put the rest of the response back into the library but that feels ... unwieldy?

**Comment by jonas-schievink** [comment]
Hmm, right... I still think that cookie support should not be part of this library (cookies aren't part of XMLRPC, although they are of course frequently used with any service that uses HTTP). You'll probably end up with a wrapper that does the cookie management anyways, so maybe the `write_as_xml` isn't too far off from what you'll end up with even if we made the API nicer.

We can try to add a nice API that allows access to the returned `Response`, but I'm not sure how that would look (just returning a tuple is possible but not very nice).

**Comment by Nervengift** [comment]
I think the easiest to use option would be to allow more than just `hyper::Client` (via trait or enum?) so that one can build a wrapped version that stores cookies between transactions

**Comment by jonas-schievink** [comment]
Yeah, defining some sort of trait sounds like a good idea. Just have to make sure it doesn't make the API too clunky or hard to understand.

**Comment by Nervengift** [comment]
(actually I think `hyper::Client` should just provide optional cookie support, but I'm not if that will happen)

**Comment by Nervengift** [comment]
If we provide a impl of that trait for `hyper::Client` it should work as before, shouldn't it?

**Comment by jonas-schievink** [comment]
Yeah, built-in cookie support in hyper would be nice. Apparently reqwest is going to get cookie support at some point, so that's another option.

> If we provide a impl of that trait for hyper::Client it should work as before, shouldn't it?

Yes, sure. I just want to make sure the API is easy to use, but this seems like a useful thing to do. I'll have to put some work into better docs anyway :)

**Comment by Nervengift** [comment]
I would suggest something like
```
pub trait MyClient {
    fn post(&self, uri: Url) -> RequestBuilder;
}

impl MyClient for hyper::Client {
    fn post(&self, uri: Url) -> RequestBuilder {
        hyper::Client::post(self, uri)
    }
}

 pub fn call(&self, client: &MyClient, url: &str) -> RequestResult {
...
```

Which seems to not break existing code since calling `call` with a `hyper::Client` is still possible.

If thats an aceptable option for you and you have a better name for the trait then I'd try to do a pull request for this and maybe an example/doc entry for using it to keep cookies

**Comment by jonas-schievink** [comment]
Looks like a step in the right direction.

I'd prefer to make `call` generic over any `T: MyClient` instead of taking a trait object (not that it matters much for performance, but generics are usually used by default).

Also `MyClient` should be renamed to more accurately reflect its use, maybe `HttpClient` or just `Client` would be a better name. `Client` does collide with hyper's `Client` type, so `HttpClient` might be better.

The rest sounds good to me, thanks for putting this together!

**Comment by Nervengift** [comment]
Next problem: this seems to work for setting the cookie before sending, but since the actual sending ist done by the `hyper::client:RequestBuilder` which only stores a reference to the original `hyper::Client`, we'll have to make the builder wrappable as well

\*gets the yak shaver\*

**Comment by jonas-schievink** [comment]
We should be able do make the trait mostly independent of hyper so that it performs the request and returns the server's response. That way your wrapper can do whatever it needs to do before or after the request happens.

**Comment by Nervengift** [comment]
I'll still need to wrap both `hyper::Client` and `hyper::client::RequestBuilder`:

The client is needed at least for storing the cookie somewhere

The request builder to access the response headers since the `.send(...)` method is part of the builder and there is no "give me the request, I'll send it myself" option

Manipulating the request headers can be done both in request builder and client (which constructs the request builder)

**Comment by jonas-schievink** [comment]
I think we mean different things. What about using this trait definition instead:

```rust
pub trait HttpClient {
    type Response: Read;

    fn request(&self, url: &str, body: &str) -> Self::Response;
}
```

This way, your custom client can inspect the server's response before handing it back to xmlrpc, and you can set your cookies before sending the request.

**Comment by Nervengift** [comment]
I had discarded a similar idea because it seemed to become really complicated since it is rather far from the approach in `hyper` but now that I think about it again, it should be doable.

I guess one problem will be that building a custom client will become more complicated. But most of that work will be done in the `impl HttpClient for hyper::Client` so that should be reusable.

I'll try to come up with something

**Comment by jonas-schievink** [comment]
A simpler approach to this was merged with #21, so I'll close this. If it's not flexible enough, feel free to open a new issue.
